In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# 데이터 불러오기
train_cn7 = pd.read_csv(r"C:\Users\Administrator\Desktop\kamp\1. 사출성형기 AI 데이터셋\1. 사출성형기 AI 데이터셋\moldset_labeled_cn7.csv")
train_rg3 = pd.read_csv(r"C:\Users\Administrator\Desktop\kamp\1. 사출성형기 AI 데이터셋\1. 사출성형기 AI 데이터셋\moldset_labeled_rg3.csv")

In [13]:
# 모든 값이 같은 컬럼 확인
train_cn7_constant = train_cn7.columns[train_cn7.nunique() == 1]
test_cn7_constant = train_rg3.columns[train_cn7.nunique() == 1]


train_cn7_constant, test_cn7_constant

(Index(['Clamp_Open_Position'], dtype='str'),
 Index(['Clamp_Open_Position'], dtype='str'))

In [14]:
train_cn7 = train_cn7.drop(['Clamp_Open_Position', 'Unnamed: 0'], axis=1)
train_rg3 = train_rg3.drop(['Clamp_Open_Position', 'Unnamed: 0'], axis=1)

In [18]:
train_cn7.duplicated().sum()

np.int64(594)

In [19]:
train_cn7[train_cn7.duplicated(keep=False)]

,PassOrFail,Injection_Time,Filling_Time,Plasticizing_Time,Cycle_Time,Clamp_Close_Time,Cushion_Position,Plasticizing_Position,Max_Injection_Speed,Max_Screw_RPM,...,Average_Back_Pressure,Barrel_Temperature_1,Barrel_Temperature_2,Barrel_Temperature_3,Barrel_Temperature_4,Barrel_Temperature_5,Barrel_Temperature_6,Hopper_Temperature,Mold_Temperature_3,Mold_Temperature_4
1,0,2.080358,1.894372,-0.527100,1.886261,0.978580,-0.711164,1.471608,-1.347095,0.049553,...,-0.533773,0.056153,0.622557,0.914107,0.504051,-0.915098,-2.263700,-0.874018,1.023094,1.337339
2,0,2.080358,1.894372,-0.527100,1.886261,0.978580,-0.711164,1.471608,-1.347095,0.049553,...,-0.533773,0.056153,0.622557,0.914107,0.504051,-0.915098,-2.263700,-0.874018,1.023094,1.337339
3,0,1.679123,1.894372,-0.527100,1.377857,0.978580,1.202502,1.471608,-1.347095,1.740153,...,-0.533773,0.729073,-0.765355,1.231418,0.886706,-0.299626,0.013158,-1.176164,1.166162,1.337339
4,0,1.679123,1.894372,-0.527100,1.377857,0.978580,1.202502,1.471608,-1.347095,1.740153,...,-0.533773,0.729073,-0.765355,1.231418,0.886706,-0.299626,0.013158,-1.176164,1.166162,1.337339
5,0,1.277888,1.088442,-0.576184,1.886261,0.978580,-0.711164,1.471608,-0.763018,0.894853,...,-0.152634,0.280414,0.067358,0.596893,1.269321,0.008133,0.013158,-0.571865,1.166162,1.337339
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1206,0,-1.931915,-1.732320,5.608208,-1.164166,-0.902466,-0.711164,-1.185844,2.352070,0.894853,...,-0.914898,0.280414,0.067358,1.548728,1.014257,0.623605,0.923887,0.737456,0.093154,-0.155625
1207,0,-4.740484,-4.553063,9.829303,-1.672571,-0.902466,3.116169,-1.487849,5.467151,1.740153,...,-3.201688,-0.841164,1.177755,-0.989563,1.779489,0.315892,0.013158,1.442471,0.093154,-0.205391
1208,0,-4.740484,-4.553063,9.829303,-1.672571,-0.902466,3.116169,-1.487849,5.467151,1.740153,...,-3.201688,-0.841164,1.177755,-0.989563,1.779489,0.315892,0.013158,1.442471,0.093154,-0.205391
1209,0,-7.549091,-7.776799,20.038456,-3.706092,-0.902466,6.943503,-2.876962,8.776929,3.430752,...,-6.631852,0.056153,-0.765355,0.914107,1.651976,-0.299626,2.290016,1.442471,0.093154,-0.205391


In [20]:
train_cn7[train_cn7.duplicated(keep=False)].drop_duplicates().shape[0]

594

In [21]:
train_cn7.value_counts().head(20)

PassOrFail  Injection_Time  Filling_Time  Plasticizing_Time  Cycle_Time  Clamp_Close_Time  Cushion_Position  Plasticizing_Position  Max_Injection_Speed  Max_Screw_RPM  Average_Screw_RPM  Max_Injection_Pressure  Max_Switch_Over_Pressure  Max_Back_Pressure  Average_Back_Pressure  Barrel_Temperature_1  Barrel_Temperature_2  Barrel_Temperature_3  Barrel_Temperature_4  Barrel_Temperature_5  Barrel_Temperature_6  Hopper_Temperature  Mold_Temperature_3  Mold_Temperature_4
0           2.080358        1.894372      -0.527100          1.886261    0.978580          -0.711164         1.471608               -1.347095             0.049553      -0.132022           0.635210                1.573231                 -0.740752          -0.533773               0.056153              0.622557              0.914107              0.504051             -0.915098             -2.263700             -0.874018           1.023094            1.337339              2
            1.679123        1.894372      -0.527100    

In [22]:
train_cn7.duplicated().mean() * 100

np.float64(49.05037159372419)

In [24]:
train_rg3[train_rg3.duplicated(keep=False)].drop_duplicates().shape[0]

566

In [25]:
train_rg3.value_counts().head(20)

PassOrFail  Injection_Time  Filling_Time  Plasticizing_Time  Cycle_Time  Clamp_Close_Time  Cushion_Position  Plasticizing_Position  Max_Injection_Speed  Max_Screw_RPM  Average_Screw_RPM  Max_Injection_Pressure  Max_Switch_Over_Pressure  Max_Back_Pressure  Average_Back_Pressure  Barrel_Temperature_1  Barrel_Temperature_2  Barrel_Temperature_3  Barrel_Temperature_4  Barrel_Temperature_5  Barrel_Temperature_6  Hopper_Temperature  Mold_Temperature_3  Mold_Temperature_4
0           -0.029099       -1.192531     1.977737            0.249511   1.033078          2.867608          -1.755816              2.187583              0.677109      -0.421798          -2.077961               -1.184902                 -1.330250          -1.306878               0.096629             -0.496175             -0.002081              0.047749              0.282727             -0.354184              0.183232           -1.393485           -1.576042             2
                                          1.528778     

In [23]:
train_rg3.duplicated().mean() * 100

np.float64(47.88494077834179)

In [26]:
train_cn7[
    (train_cn7['PassOrFail'] == 1) &
    (train_cn7.duplicated(keep=False))
].shape[0]

6

In [27]:
train_rg3[
    (train_rg3['PassOrFail'] == 1) &
    (train_rg3.duplicated(keep=False))
].shape[0]

0

In [28]:
# 동일 X + 다른 Target 확인
X_cols = train_cn7.columns.drop('PassOrFail')

same_x_diff_target = train_cn7[
    train_cn7.groupby(list(X_cols))['PassOrFail']
             .transform('nunique') > 1
]

same_x_diff_target

,PassOrFail,Injection_Time,Filling_Time,Plasticizing_Time,Cycle_Time,Clamp_Close_Time,Cushion_Position,Plasticizing_Position,Max_Injection_Speed,Max_Screw_RPM,...,Average_Back_Pressure,Barrel_Temperature_1,Barrel_Temperature_2,Barrel_Temperature_3,Barrel_Temperature_4,Barrel_Temperature_5,Barrel_Temperature_6,Hopper_Temperature,Mold_Temperature_3,Mold_Temperature_4
47,1,1.277888,1.491397,-0.625267,0.361047,0.978580,-0.711164,1.471608,-1.152398,-1.641030,...,-0.152634,-0.168176,-0.210156,0.279582,1.269321,0.623605,0.923887,-0.269711,1.166162,1.287574
48,0,1.277888,1.491397,-0.625267,0.361047,0.978580,-0.711164,1.471608,-1.152398,-1.641030,...,-0.152634,-0.168176,-0.210156,0.279582,1.269321,0.623605,0.923887,-0.269711,1.166162,1.287574
79,0,2.080358,1.894372,-0.478017,0.361047,0.978580,-0.711164,1.471608,-1.347095,0.049553,...,-0.533773,-0.392505,0.900156,-0.037728,-0.388733,1.546836,0.923887,-1.276879,1.309231,1.387105
80,1,2.080358,1.894372,-0.478017,0.361047,0.978580,-0.711164,1.471608,-1.347095,0.049553,...,-0.533773,-0.392505,0.900156,-0.037728,-0.388733,1.546836,0.923887,-1.276879,1.309231,1.387105
95,1,1.679123,1.894372,-0.330766,0.361047,0.978580,1.202502,1.471608,-1.347095,-0.795730,...,-0.914898,-0.168176,2.843181,0.596893,-0.516285,-2.761513,1.834616,0.435303,1.309231,1.387105
96,0,1.679123,1.894372,-0.330766,0.361047,0.978580,1.202502,1.471608,-1.347095,-0.795730,...,-0.914898,-0.168176,2.843181,0.596893,-0.516285,-2.761513,1.834616,0.435303,1.309231,1.387105
99,1,1.679123,1.491397,-0.428933,1.377857,0.978580,-0.711164,1.471608,-0.957708,-1.641030,...,-0.533773,-1.065425,0.344957,-1.941496,0.631602,0.315892,-1.352970,-0.773295,1.237697,1.387105
100,0,1.679123,1.491397,-0.428933,1.377857,0.978580,-0.711164,1.471608,-0.957708,-1.641030,...,-0.533773,-1.065425,0.344957,-1.941496,0.631602,0.315892,-1.352970,-0.773295,1.237697,1.387105
101,1,1.277888,1.491397,-0.428933,1.377857,0.978580,-0.711164,1.471608,-0.957708,0.894853,...,-0.533773,1.177731,-1.320553,0.279582,1.141770,0.315892,-1.352970,-1.075448,1.309231,1.387105
102,0,1.277888,1.491397,-0.428933,1.377857,0.978580,-0.711164,1.471608,-0.957708,0.894853,...,-0.533773,1.177731,-1.320553,0.279582,1.141770,0.315892,-1.352970,-1.075448,1.309231,1.387105


In [29]:
same_x_diff_target.shape[0]

22

In [30]:
same_x_diff_target[X_cols].drop_duplicates().shape[0]

11

동일한 관측 공정조건(X)에서도 정상과 불량이 동시에 발생함.
현재 데이터에 포함되지 않은 미관측 공정변수, 공정 자체의 변동성, 센서 측정 오차 또는 라벨 노이즈 가능성을 고려할 필요가 있음.

In [46]:
# Target 제외
X_cols_cn7 = train_cn7.columns.drop('PassOrFail')
X_cols_rg3 = train_rg3.columns.drop('PassOrFail')


In [48]:
# 하나의 변수라도 |z| > 3인 행
outlier_mask_cn7 = (train_cn7[X_cols_cn7].abs() > 3).any(axis=1)
outlier_mask_rg3 = (train_rg3[X_cols_rg3].abs() > 3).any(axis=1)


In [55]:
X_cols_cn7 = train_cn7.columns.drop('PassOrFail')

same_x_diff_target_cn7 = train_cn7[
    train_cn7.groupby(list(X_cols_cn7))['PassOrFail']
             .transform('nunique') > 1
]

same_x_diff_target_cn7

same_x_diff_target_cn7.shape[0]

22

In [53]:
X_cols_rg3 = train_rg3.columns.drop('PassOrFail')

same_x_diff_target_rg3 = train_rg3[
    train_rg3.groupby(list(X_cols_rg3))['PassOrFail']
             .transform('nunique') > 1
]

same_x_diff_target_rg3

same_x_diff_target_rg3.shape[0]

50

전체 데이터의 약 **3.96%**에서 하나 이상의 공정변수가 |z| > 3 범위를 벗어나는 이상치 후보가 확인되었다. 비율이 과도하게 높지는 않으므로 일괄 제거하지 않고, 해당 이상치가 불량(PassOrFail=1)과 관련된 공정 이상 신호인지 추가 분석한다.

In [49]:
print("이상치 있음")
print(train_cn7.loc[outlier_mask_cn7, 'PassOrFail'].value_counts(normalize=True) * 100)

print("\n이상치 없음")
print(train_cn7.loc[~outlier_mask_cn7, 'PassOrFail'].value_counts(normalize=True) * 100)

이상치 있음
PassOrFail
0    87.5
1    12.5
Name: proportion, dtype: float64

이상치 없음
PassOrFail
0    99.05417
1     0.94583
Name: proportion, dtype: float64


In [52]:
print("이상치 있음")
print(train_rg3.loc[outlier_mask_rg3, 'PassOrFail'].value_counts(normalize=True) * 100)

print("\n이상치 없음")
print(train_rg3.loc[~outlier_mask_rg3, 'PassOrFail'].value_counts(normalize=True) * 100)

이상치 있음
PassOrFail
0    100.0
Name: proportion, dtype: float64

이상치 없음
PassOrFail
0    97.779751
1     2.220249
Name: proportion, dtype: float64


In [58]:
same_x_diff_target_cn7[X_cols_cn7].drop_duplicates().shape[0]


11

In [57]:
same_x_diff_target_rg3[X_cols_rg3].drop_duplicates().shape[0]

25

In [59]:
conflict_fail_cn7 = (
    same_x_diff_target_cn7['PassOrFail'] == 1
).sum()

total_fail_cn7 = (
    train_cn7['PassOrFail'] == 1
).sum()

print("충돌 데이터 내 불량:", conflict_fail_cn7)
print("전체 불량:", total_fail_cn7)
print(
    "전체 불량 중 충돌 데이터 비율:",
    conflict_fail_cn7 / total_fail_cn7 * 100
)

충돌 데이터 내 불량: 11
전체 불량: 17
전체 불량 중 충돌 데이터 비율: 64.70588235294117


In [60]:
conflict_fail_rg3 = (
    same_x_diff_target_rg3['PassOrFail'] == 1
).sum()

total_fail_rg3 = (
    train_rg3['PassOrFail'] == 1
).sum()

print("충돌 데이터 내 불량:", conflict_fail_rg3)
print("전체 불량:", total_fail_rg3)
print(
    "전체 불량 중 충돌 데이터 비율:",
    conflict_fail_rg3 / total_fail_rg3 * 100
)

충돌 데이터 내 불량: 25
전체 불량: 25
전체 불량 중 충돌 데이터 비율: 100.0


In [61]:
same_x_diff_target_rg3['PassOrFail'].value_counts()

PassOrFail
1    25
0    25
Name: count, dtype: int64

CN7은 극단적인 공정값에서 불량률이 약 13배 증가하여 이상치 자체가 불량 신호로 활용될 가능성이 확인되었다. 반면 RG3에서는 이상치 구간에서 불량이 발생하지 않았으며, 전체 불량 25건이 모두 동일 공정조건에서 정상과 함께 관측되었다. 따라서 두 제품의 불량 발생 구조가 상이하며, CN7은 공정값의 극단적 변화, RG3는 미관측 요인 또는 공정변수 간 복합적인 관계를 추가적으로 고려할 필요가 있다.

In [62]:
# 원본 변수는 그대로 유지
X_cols_cn7 = train_cn7.columns.drop('PassOrFail')

# 하나라도 |z| > 3이면 1
train_cn7['is_outlier'] = (
    train_cn7[X_cols_cn7].abs() > 3
).any(axis=1).astype(int)

# 한 행에서 이상치인 공정변수가 몇 개인지
train_cn7['outlier_count'] = (
    train_cn7[X_cols_cn7].abs() > 3
).sum(axis=1)

In [63]:
for col in X_cols_cn7:
    train_cn7[f'{col}_outlier'] = (
        train_cn7[col].abs() > 3
    ).astype(int)

In [64]:
X_cols_rg3 = train_rg3.columns.drop('PassOrFail')

이상치는 일괄 삭제하지 않는다. Standard Scaling 기준 |z| > 3을 이상치 후보로 정의하여 Target과의 관계를 분석하였다. CN7에서는 이상치 데이터의 불량률이 12.5%로 비이상치 데이터의 0.95%보다 크게 높아, 이상치를 공정 이상 신호로 판단하고 원본 값을 유지하면서 is_outlier, outlier_count 등의 파생변수로 활용한다. 반면 RG3에서는 이상치 구간에서 불량이 발생하지 않아 이상치를 제거하지 않고 원본 값 그대로 사용하며, 이상치 관련 파생변수의 효과는 모델 성능 비교를 통해 판단한다.

In [65]:
import numpy as np
import pandas as pd

from sklearn.model_selection import StratifiedGroupKFold
from sklearn.linear_model import LogisticRegression
from sklearn.dummy import DummyClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    average_precision_score,
    confusion_matrix
)

In [66]:
from sklearn.model_selection import train_test_split

X_cn7 = train_cn7.drop(columns='PassOrFail')
y_cn7 = train_cn7['PassOrFail']

X_train_cn7, X_test_cn7, y_train_cn7, y_test_cn7 = train_test_split(
    X_cn7,
    y_cn7,
    test_size=0.2,
    stratify=y_cn7,
    random_state=42
)

In [67]:
X_rg3 = train_rg3.drop(columns='PassOrFail')
y_rg3 = train_rg3['PassOrFail']

X_train_rg3, X_test_rg3, y_train_rg3, y_test_rg3 = train_test_split(
    X_rg3,
    y_rg3,
    test_size=0.2,
    stratify=y_rg3,
    random_state=42
)

In [71]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    average_precision_score,
    confusion_matrix,
    classification_report
)

# X / y 분리
X_cn7 = train_cn7.drop(columns='PassOrFail')
y_cn7 = train_cn7['PassOrFail']

# Stratified Split
X_train_cn7, X_test_cn7, y_train_cn7, y_test_cn7 = train_test_split(
    X_cn7,
    y_cn7,
    test_size=0.2,
    stratify=y_cn7,
    random_state=42
)

# Logistic Regression
logistic_cn7 = LogisticRegression(
    class_weight='balanced',
    max_iter=1000,
    random_state=42
)

logistic_cn7.fit(X_train_cn7, y_train_cn7)

# 예측
pred_cn7 = logistic_cn7.predict(X_test_cn7)
proba_cn7 = logistic_cn7.predict_proba(X_test_cn7)[:, 1]

In [72]:
print("[CN7 Logistic Regression]")

print("Accuracy :", accuracy_score(y_test_cn7, pred_cn7))
print("Precision:", precision_score(y_test_cn7, pred_cn7, zero_division=0))
print("Recall   :", recall_score(y_test_cn7, pred_cn7, zero_division=0))
print("F1       :", f1_score(y_test_cn7, pred_cn7, zero_division=0))
print("PR-AUC   :", average_precision_score(y_test_cn7, proba_cn7))

print("\nConfusion Matrix")
print(confusion_matrix(y_test_cn7, pred_cn7))

print("\nClassification Report")
print(classification_report(y_test_cn7, pred_cn7, digits=4))

[CN7 Logistic Regression]
Accuracy : 0.9629629629629629
Precision: 0.125
Recall   : 0.3333333333333333
F1       : 0.18181818181818182
PR-AUC   : 0.1535885167464115

Confusion Matrix
[[233   7]
 [  2   1]]

Classification Report
              precision    recall  f1-score   support

           0     0.9915    0.9708    0.9811       240
           1     0.1250    0.3333    0.1818         3

    accuracy                         0.9630       243
   macro avg     0.5582    0.6521    0.5814       243
weighted avg     0.9808    0.9630    0.9712       243



In [74]:
# X / y 분리
X_rg3 = train_rg3.drop(columns='PassOrFail')
y_rg3 = train_rg3['PassOrFail']

# Stratified Split
X_train_rg3, X_test_rg3, y_train_rg3, y_test_rg3 = train_test_split(
    X_rg3,
    y_rg3,
    test_size=0.2,
    stratify=y_rg3,
    random_state=42
)

# Logistic Regression
logistic_rg3 = LogisticRegression(
    class_weight='balanced',
    max_iter=1000,
    random_state=42
)

logistic_rg3.fit(X_train_rg3, y_train_rg3)

# 예측
pred_rg3 = logistic_rg3.predict(X_test_rg3)
proba_rg3 = logistic_rg3.predict_proba(X_test_rg3)[:, 1]

In [75]:
print("[RG3 Logistic Regression]")

print("Accuracy :", accuracy_score(y_test_rg3, pred_rg3))
print("Precision:", precision_score(y_test_rg3, pred_rg3, zero_division=0))
print("Recall   :", recall_score(y_test_rg3, pred_rg3, zero_division=0))
print("F1       :", f1_score(y_test_rg3, pred_rg3, zero_division=0))
print("PR-AUC   :", average_precision_score(y_test_rg3, proba_rg3))

print("\nConfusion Matrix")
print(confusion_matrix(y_test_rg3, pred_rg3))

print("\nClassification Report")
print(classification_report(y_test_rg3, pred_rg3, digits=4))

[RG3 Logistic Regression]
Accuracy : 0.6877637130801688
Precision: 0.0273972602739726
Recall   : 0.4
F1       : 0.05128205128205128
PR-AUC   : 0.04721508939561678

Confusion Matrix
[[161  71]
 [  3   2]]

Classification Report
              precision    recall  f1-score   support

           0     0.9817    0.6940    0.8131       232
           1     0.0274    0.4000    0.0513         5

    accuracy                         0.6878       237
   macro avg     0.5046    0.5470    0.4322       237
weighted avg     0.9616    0.6878    0.7971       237



In [76]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    average_precision_score,
    confusion_matrix,
    classification_report
)

rf_cn7 = RandomForestClassifier(
    n_estimators=300,
    class_weight='balanced',
    random_state=42,
    n_jobs=-1
)

rf_cn7.fit(X_train_cn7, y_train_cn7)

pred_rf_cn7 = rf_cn7.predict(X_test_cn7)
proba_rf_cn7 = rf_cn7.predict_proba(X_test_cn7)[:, 1]

print("[CN7 Random Forest]")

print("Accuracy :", accuracy_score(y_test_cn7, pred_rf_cn7))
print("Precision:", precision_score(y_test_cn7, pred_rf_cn7, zero_division=0))
print("Recall   :", recall_score(y_test_cn7, pred_rf_cn7, zero_division=0))
print("F1       :", f1_score(y_test_cn7, pred_rf_cn7, zero_division=0))
print("PR-AUC   :", average_precision_score(y_test_cn7, proba_rf_cn7))

print("\nConfusion Matrix")
print(confusion_matrix(y_test_cn7, pred_rf_cn7))

print("\nClassification Report")
print(classification_report(y_test_cn7, pred_rf_cn7, digits=4))

[CN7 Random Forest]
Accuracy : 0.9835390946502057
Precision: 0.0
Recall   : 0.0
F1       : 0.0
PR-AUC   : 0.14699074074074073

Confusion Matrix
[[239   1]
 [  3   0]]

Classification Report
              precision    recall  f1-score   support

           0     0.9876    0.9958    0.9917       240
           1     0.0000    0.0000    0.0000         3

    accuracy                         0.9835       243
   macro avg     0.4938    0.4979    0.4959       243
weighted avg     0.9754    0.9835    0.9795       243



In [77]:
rf_rg3 = RandomForestClassifier(
    n_estimators=300,
    class_weight='balanced',
    random_state=42,
    n_jobs=-1
)

rf_rg3.fit(X_train_rg3, y_train_rg3)

pred_rf_rg3 = rf_rg3.predict(X_test_rg3)
proba_rf_rg3 = rf_rg3.predict_proba(X_test_rg3)[:, 1]

print("[RG3 Random Forest]")

print("Accuracy :", accuracy_score(y_test_rg3, pred_rf_rg3))
print("Precision:", precision_score(y_test_rg3, pred_rf_rg3, zero_division=0))
print("Recall   :", recall_score(y_test_rg3, pred_rf_rg3, zero_division=0))
print("F1       :", f1_score(y_test_rg3, pred_rf_rg3, zero_division=0))
print("PR-AUC   :", average_precision_score(y_test_rg3, proba_rf_rg3))

print("\nConfusion Matrix")
print(confusion_matrix(y_test_rg3, pred_rf_rg3))

print("\nClassification Report")
print(classification_report(y_test_rg3, pred_rf_rg3, digits=4))

[RG3 Random Forest]
Accuracy : 0.9535864978902954
Precision: 0.0
Recall   : 0.0
F1       : 0.0
PR-AUC   : 0.02461994507282751

Confusion Matrix
[[226   6]
 [  5   0]]

Classification Report
              precision    recall  f1-score   support

           0     0.9784    0.9741    0.9762       232
           1     0.0000    0.0000    0.0000         5

    accuracy                         0.9536       237
   macro avg     0.4892    0.4871    0.4881       237
weighted avg     0.9577    0.9536    0.9556       237



In [78]:
comparison = pd.DataFrame({
    'Dataset': ['CN7', 'CN7', 'RG3', 'RG3'],
    'Model': ['Logistic', 'RandomForest', 'Logistic', 'RandomForest'],
    'Precision': [
        precision_score(y_test_cn7, pred_cn7, zero_division=0),
        precision_score(y_test_cn7, pred_rf_cn7, zero_division=0),
        precision_score(y_test_rg3, pred_rg3, zero_division=0),
        precision_score(y_test_rg3, pred_rf_rg3, zero_division=0)
    ],
    'Recall': [
        recall_score(y_test_cn7, pred_cn7, zero_division=0),
        recall_score(y_test_cn7, pred_rf_cn7, zero_division=0),
        recall_score(y_test_rg3, pred_rg3, zero_division=0),
        recall_score(y_test_rg3, pred_rf_rg3, zero_division=0)
    ],
    'F1': [
        f1_score(y_test_cn7, pred_cn7, zero_division=0),
        f1_score(y_test_cn7, pred_rf_cn7, zero_division=0),
        f1_score(y_test_rg3, pred_rg3, zero_division=0),
        f1_score(y_test_rg3, pred_rf_rg3, zero_division=0)
    ],
    'PR_AUC': [
        average_precision_score(y_test_cn7, proba_cn7),
        average_precision_score(y_test_cn7, proba_rf_cn7),
        average_precision_score(y_test_rg3, proba_rg3),
        average_precision_score(y_test_rg3, proba_rf_rg3)
    ]
})

comparison

,Dataset,Model,Precision,Recall,F1,PR_AUC
0,CN7,Logistic,0.125000,0.333333,0.181818,0.153589
1,CN7,RandomForest,0.000000,0.000000,0.000000,0.146991
2,RG3,Logistic,0.027397,0.400000,0.051282,0.047215
3,RG3,RandomForest,0.000000,0.000000,0.000000,0.024620


In [79]:
# CN7 RandomForest 예측 확률 범위
print("최소:", proba_rf_cn7.min())
print("최대:", proba_rf_cn7.max())

# 높은 확률 순으로 확인
pd.Series(proba_rf_cn7).sort_values(ascending=False).head(20)

최소: 0.0
최대: 1.0


233    1.000000
68     0.257077
13     0.214311
56     0.180681
180    0.134743
200    0.132103
116    0.079050
157    0.075964
127    0.069159
147    0.065873
115    0.065735
190    0.065735
207    0.065352
76     0.062608
99     0.062399
117    0.062399
111    0.062252
208    0.059190
124    0.046269
216    0.046269
dtype: float64

지금 단계의 결론

CN7

Logistic과 RandomForest 모두 어느 정도 불량 순위화 신호는 있음. 현재 기본 threshold 기준으로는 Logistic이 우세하며, RandomForest는 확률 threshold 조정 가능성을 추가 확인할 필요가 있음.

RG3

Logistic도 낮은 성능을 보이며 RandomForest는 PR-AUC가 약 0.025로 거의 클래스 비율 수준이다. 앞서 확인한 불량 25건 전부 동일 X + 다른 Target 구조 때문에 현재 제공된 Feature만으로 분류하기 어려운 것으로 보인다.

In [81]:
from xgboost import XGBClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    average_precision_score,
    confusion_matrix,
    classification_report
)

# 클래스 불균형 비율
scale_pos_weight_cn7 = (
    (y_train_cn7 == 0).sum() /
    (y_train_cn7 == 1).sum()
)

print("scale_pos_weight:", scale_pos_weight_cn7)

xgb_cn7 = XGBClassifier(
    n_estimators=300,
    max_depth=3,
    learning_rate=0.05,
    subsample=0.8,
    colsample_bytree=0.8,
    scale_pos_weight=scale_pos_weight_cn7,
    eval_metric='logloss',
    random_state=42,
    n_jobs=-1
)

xgb_cn7.fit(X_train_cn7, y_train_cn7)

pred_xgb_cn7 = xgb_cn7.predict(X_test_cn7)
proba_xgb_cn7 = xgb_cn7.predict_proba(X_test_cn7)[:, 1]

scale_pos_weight: 68.14285714285714


In [82]:
print("[CN7 XGBoost]")

print("Accuracy :", accuracy_score(y_test_cn7, pred_xgb_cn7))
print("Precision:", precision_score(y_test_cn7, pred_xgb_cn7, zero_division=0))
print("Recall   :", recall_score(y_test_cn7, pred_xgb_cn7, zero_division=0))
print("F1       :", f1_score(y_test_cn7, pred_xgb_cn7, zero_division=0))
print("PR-AUC   :", average_precision_score(y_test_cn7, proba_xgb_cn7))

print("\nConfusion Matrix")
print(confusion_matrix(y_test_cn7, pred_xgb_cn7))

print("\nClassification Report")
print(classification_report(y_test_cn7, pred_xgb_cn7, digits=4))

[CN7 XGBoost]
Accuracy : 0.9835390946502057
Precision: 0.0
Recall   : 0.0
F1       : 0.0
PR-AUC   : 0.3416666666666667

Confusion Matrix
[[239   1]
 [  3   0]]

Classification Report
              precision    recall  f1-score   support

           0     0.9876    0.9958    0.9917       240
           1     0.0000    0.0000    0.0000         3

    accuracy                         0.9835       243
   macro avg     0.4938    0.4979    0.4959       243
weighted avg     0.9754    0.9835    0.9795       243



In [83]:
scale_pos_weight_rg3 = (
    (y_train_rg3 == 0).sum() /
    (y_train_rg3 == 1).sum()
)

print("scale_pos_weight:", scale_pos_weight_rg3)

xgb_rg3 = XGBClassifier(
    n_estimators=300,
    max_depth=3,
    learning_rate=0.05,
    subsample=0.8,
    colsample_bytree=0.8,
    scale_pos_weight=scale_pos_weight_rg3,
    eval_metric='logloss',
    random_state=42,
    n_jobs=-1
)

xgb_rg3.fit(X_train_rg3, y_train_rg3)

pred_xgb_rg3 = xgb_rg3.predict(X_test_rg3)
proba_xgb_rg3 = xgb_rg3.predict_proba(X_test_rg3)[:, 1]

print("[RG3 XGBoost]")

print("Accuracy :", accuracy_score(y_test_rg3, pred_xgb_rg3))
print("Precision:", precision_score(y_test_rg3, pred_xgb_rg3, zero_division=0))
print("Recall   :", recall_score(y_test_rg3, pred_xgb_rg3, zero_division=0))
print("F1       :", f1_score(y_test_rg3, pred_xgb_rg3, zero_division=0))
print("PR-AUC   :", average_precision_score(y_test_rg3, proba_xgb_rg3))

print("\nConfusion Matrix")
print(confusion_matrix(y_test_rg3, pred_xgb_rg3))

print("\nClassification Report")
print(classification_report(y_test_rg3, pred_xgb_rg3, digits=4))

scale_pos_weight: 46.25
[RG3 XGBoost]
Accuracy : 0.9493670886075949
Precision: 0.0
Recall   : 0.0
F1       : 0.0
PR-AUC   : 0.033485134027528646

Confusion Matrix
[[225   7]
 [  5   0]]

Classification Report
              precision    recall  f1-score   support

           0     0.9783    0.9698    0.9740       232
           1     0.0000    0.0000    0.0000         5

    accuracy                         0.9494       237
   macro avg     0.4891    0.4849    0.4870       237
weighted avg     0.9576    0.9494    0.9535       237



In [84]:
comparison = pd.DataFrame({
    'Dataset': [
        'CN7', 'CN7', 'CN7',
        'RG3', 'RG3', 'RG3'
    ],
    
    'Model': [
        'Logistic',
        'RandomForest',
        'XGBoost',
        'Logistic',
        'RandomForest',
        'XGBoost'
    ],

    'Precision': [
        precision_score(y_test_cn7, pred_cn7, zero_division=0),
        precision_score(y_test_cn7, pred_rf_cn7, zero_division=0),
        precision_score(y_test_cn7, pred_xgb_cn7, zero_division=0),

        precision_score(y_test_rg3, pred_rg3, zero_division=0),
        precision_score(y_test_rg3, pred_rf_rg3, zero_division=0),
        precision_score(y_test_rg3, pred_xgb_rg3, zero_division=0)
    ],

    'Recall': [
        recall_score(y_test_cn7, pred_cn7, zero_division=0),
        recall_score(y_test_cn7, pred_rf_cn7, zero_division=0),
        recall_score(y_test_cn7, pred_xgb_cn7, zero_division=0),

        recall_score(y_test_rg3, pred_rg3, zero_division=0),
        recall_score(y_test_rg3, pred_rf_rg3, zero_division=0),
        recall_score(y_test_rg3, pred_xgb_rg3, zero_division=0)
    ],

    'F1': [
        f1_score(y_test_cn7, pred_cn7, zero_division=0),
        f1_score(y_test_cn7, pred_rf_cn7, zero_division=0),
        f1_score(y_test_cn7, pred_xgb_cn7, zero_division=0),

        f1_score(y_test_rg3, pred_rg3, zero_division=0),
        f1_score(y_test_rg3, pred_rf_rg3, zero_division=0),
        f1_score(y_test_rg3, pred_xgb_rg3, zero_division=0)
    ],

    'PR_AUC': [
        average_precision_score(y_test_cn7, proba_cn7),
        average_precision_score(y_test_cn7, proba_rf_cn7),
        average_precision_score(y_test_cn7, proba_xgb_cn7),

        average_precision_score(y_test_rg3, proba_rg3),
        average_precision_score(y_test_rg3, proba_rf_rg3),
        average_precision_score(y_test_rg3, proba_xgb_rg3)
    ]
})

comparison

,Dataset,Model,Precision,Recall,F1,PR_AUC
0,CN7,Logistic,0.125000,0.333333,0.181818,0.153589
1,CN7,RandomForest,0.000000,0.000000,0.000000,0.146991
2,CN7,XGBoost,0.000000,0.000000,0.000000,0.341667
3,RG3,Logistic,0.027397,0.400000,0.051282,0.047215
4,RG3,RandomForest,0.000000,0.000000,0.000000,0.024620
5,RG3,XGBoost,0.000000,0.000000,0.000000,0.033485


In [85]:
print("최대 확률:", proba_xgb_cn7.max())
print("최소 확률:", proba_xgb_cn7.min())

pd.DataFrame({
    '실제값': y_test_cn7.values,
    '불량확률': proba_xgb_cn7
}).sort_values(
    '불량확률',
    ascending=False
).head(20)

최대 확률: 0.99113196
최소 확률: 2.4357872e-05


,실제값,불량확률
233,0,0.991132
127,1,0.128787
13,0,0.101152
56,0,0.089307
157,1,0.049698
180,0,0.035286
68,0,0.033886
111,0,0.016329
116,0,0.015416
214,0,0.009522


In [86]:
for threshold in [0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.4, 0.5]:

    pred = (proba_xgb_cn7 >= threshold).astype(int)

    print(f"\nThreshold = {threshold}")
    print(
        "Precision:",
        round(precision_score(
            y_test_cn7, pred, zero_division=0
        ), 3)
    )
    print(
        "Recall:",
        round(recall_score(
            y_test_cn7, pred, zero_division=0
        ), 3)
    )
    print(
        "F1:",
        round(f1_score(
            y_test_cn7, pred, zero_division=0
        ), 3)
    )


Threshold = 0.05
Precision: 0.25
Recall: 0.333
F1: 0.286

Threshold = 0.1
Precision: 0.333
Recall: 0.333
F1: 0.333

Threshold = 0.15
Precision: 0.0
Recall: 0.0
F1: 0.0

Threshold = 0.2
Precision: 0.0
Recall: 0.0
F1: 0.0

Threshold = 0.25
Precision: 0.0
Recall: 0.0
F1: 0.0

Threshold = 0.3
Precision: 0.0
Recall: 0.0
F1: 0.0

Threshold = 0.4
Precision: 0.0
Recall: 0.0
F1: 0.0

Threshold = 0.5
Precision: 0.0
Recall: 0.0
F1: 0.0
